File: module/ngflink/src/main/java/me/samei/xtool/esreporter/v1/flink/GroupedMetrics.java
package me.samei.xtool.esreporter.v1.flink;

import me.samei.xtool.esreporter.v1.common.MetaData;
import me.samei.xtool.esreporter.v1.common.Value;
import org.apache.flink.metrics.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

public class GroupedMetrics {

    private static class Container {
        public final String key;
        public final String name;
        public final Metric metric;
        public final MetricGroup group;
        public Container(String key, String name, Metric metric, MetricGroup group) {
            this.key = key;
            this.name = name;
            this.metric = metric;
            this.group = group;
        }
    }

    protected final Logger logger = LoggerFactory.getLogger(getClass());

    protected Formatter formatter = new Formatter();

    protected final Map<Gauge<?>, Container> gauges = new HashMap<>();
    protected final Map<Counter, Container> counters = new HashMap<>();
    protected final Map<Histogram, Container> histograms = new HashMap<>();
    protected final Map<Meter, Container> meters = new HashMap<>();

    protected final Map<String, String> vars = new HashMap<>();
    protected final Map<String, Integer> varsCount = new HashMap<>();

    protected void addVar(String key, String val) {
        int count = varsCount.getOrDefault(key, 0);
        if (count == 0) { vars.put(key, val); }
        varsCount.put(key, count + 1);
    }

    protected void removeVar(String key) {
        int count = varsCount.getOrDefault(key, 0);
        if (count == 0) {
            varsCount.remove(key);
            vars.remove(key);
        } else {
            varsCount.put(key, count - 1);
        }
    }

    protected void addVars(MetricGroup group) {
        for (Map.Entry<String, String> item: group.getAllVariables().entrySet()) {
            addVar(item.getKey(), item.getValue());
        }
    }

    protected void removeVars(MetricGroup group) {
        for (Map.Entry<String, String> item: group.getAllVariables().entrySet()) {
            removeVar(item.getKey());
        }
    }


    public void add(String key, Metric metric, String name, MetricGroup group) {
        Container cntr = new Container(key, name, metric, group);
        if (metric instanceof Counter) {
            counters.put((Counter) metric, cntr);
        } else if (metric instanceof Gauge) {
            gauges.put((Gauge<?>) metric, cntr);
        } else if (metric instanceof Histogram) {
            histograms.put((Histogram) metric, cntr);
        } else if (metric instanceof Meter) {
            meters.put((Meter) metric, cntr);
        } else {
            logger.warn("Add, Unknown Metric Type: {}, Key: {}", metric.getClass().getName(), key);
        }
    }

    public void remove(String key, Metric metric, String name, MetricGroup group) {
        if (metric instanceof Counter) {
            counters.remove((Counter) metric);
        } else if (metric instanceof Gauge) {
            gauges.remove((Gauge) metric);
        } else if (metric instanceof Histogram) {
            histograms.remove((Histogram) metric);
        } else if (metric instanceof Meter) {
            meters.remove((Meter) metric);
        } else {
            logger.warn("Remove, Unknown Metric Type: {}, Key: {}", metric.getClass().getName(), key);
        }
    }

    protected Collection<Value> collect() {

        ArrayList<Value> temp = new ArrayList<>();

        for(Map.Entry<Gauge<?>, Container> item: gauges.entrySet()) {
            temp.add(formatter.fromGauge(item.getValue().key, item.getKey()));
        }

        for(Map.Entry<Counter, Container> item: counters.entrySet()) {
            temp.add(formatter.fromCounter(item.getValue().key, item.getKey()));
        }

        for(Map.Entry<Histogram, Container> item: histograms.entrySet()) {
            for(Value value: formatter.fromHistogram(item.getValue().key, item.getKey())) {
                temp.add(value);
            }
        }

        for(Map.Entry<Meter, Container> item: meters.entrySet()) {
            for(Value value: formatter.fromMeter(item.getValue().key, item.getKey())) {
                temp.add(value);
            }
        }

        HashMap<String, Value> result = new HashMap();
        for(Value v: temp) {
            Value already = result.get(v.key);
            if (already != null) {
                if (already.value.equals(v.value)) {
                    if (logger.isDebugEnabled()) logger.debug("Collect, Duplicated Values, Ignore New One, {}, {}", already, v);
                } else logger.warn("Collect, Duplicated Keys, Ignore New One, {}, {}", already, v);
            } else result.put(v.key, v);
        }

        return result.values();
    }

    protected Collection<Value> collectVars(MetaData meta) {
        ArrayList<Value> list = new ArrayList<>();
        for (Map.Entry<String, String> item: vars.entrySet()) {
            String key = meta.keyWith("var" + "." + item.getKey());
            Value value = formatter.fromString(key, item.getValue());
            list.add(value);
        }
        return list;
    }


}

File: module/flink14/src/main/scala/com/sameei/xtool/elasticreporter/v1/flink/lego/GroupedMetrics.scala
package com.sameei.xtool.elasticreporter.v1.flink.lego
import com.sameei.xtool.elasticreporter.v1.common.ReportContext
import org.apache.flink.metrics._
import org.slf4j.LoggerFactory
import scala.collection.JavaConverters._


class GroupedMetrics(
    override val id: String
) extends com.sameei.xtool.elasticreporter.v1.common.GroupedMetrics with Formatters {

    protected val logger = LoggerFactory getLogger id

    protected case class MetricRep(key: String, name: String, metric: Metric, group: MetricGroup) {
        val metricId = group.getMetricIdentifier(name)
    }

    protected case class VarRep(name: String, value: String, var count: Int)

    protected def empty[T] = collection.mutable.HashMap.empty[T, MetricRep]

    protected val gauges = empty[Gauge[_]]
    protected val counters = empty[Counter]
    protected val histograms = empty[Histogram]
    protected val meters = empty[Meter]

    protected val byKey = empty[String]

    protected val vars = collection.mutable.HashMap.empty[String, VarRep]

    logger.info(s"Init, Class: ${getClass.getName}")

    def count = byKey.size

    def isEmpty = byKey.isEmpty

    protected def addVar(key: String, value: String): Unit = {
        vars.get(key) match {
            case None =>
                vars.put(key, VarRep(key, value, 0))
                if (logger.isDebugEnabled()) logger.debug(s"AddVar, Key: ${key}, Val: ${value}")
            case Some(current) if current.value == value => current.count += 1
            case Some(current) =>
                logger.warn(s"AddVar, Simillar Key Different Values, Key: '${key}', Old: '${current.value}', New: '${value}', IGNORE NEW!")
        }
    }

    protected def dropVar(key: String): Unit = {
        vars.get(key) match {
            case Some(current) if current.count == 1 =>
                vars.remove(key)
                if (logger.isDebugEnabled()) logger.debug(s"AddVar, Key: ${key}, Val: ${current.value}")
            case Some(current) => current.count -= 1
            case _ =>
        }
    }

    def addMetric(key: String, name: String, metric: Metric, group: MetricGroup) = {

        byKey.get(key) match {
            case Some(v) =>
                val id = group.getMetricIdentifier(name)
                logger.warn(s"AddMetric, Key Conflict: '${key}', Old: ${v.metricId}, New: ${id}, IGNORE NEW!")

            case None =>

                val rp = MetricRep(key, name, metric, group)

                val hasBeenAdded = metric match {
                    case counter: Counter => counters.put(counter, rp); true
                    case gauge: Gauge[_] => gauges.put(gauge, rp); true
                    case histogram: Histogram => histograms.put(histogram, rp); true
                    case meter: Meter => meters.put(meter, rp); true
                    case _ => false
                }

                if (hasBeenAdded) {

                    byKey.put(key, rp)

                    val vars = group.getAllVariables.asScala.toSeq
                    vars.foreach { case (k,v) => addVar(k, v) }

                    if (logger.isDebugEnabled()){
                        val id = group.getMetricIdentifier(name)
                        logger.debug(s"AddMetric, Key: ${key}, Vars: ${vars.size}, Class: ${metric.getClass.getName}, ID: ${id},")
                    }

                } else logger.warn(s"AddMEtric, Unexpected Class: ${metric.getClass.getName}")
        }
    }

    def dropMetric(metric: Metric) : Unit = {

        val rp = metric match {
            case counter: Counter => counters.remove(counter)
            case gauge: Gauge[_] => gauges.remove(gauge)
            case histogram: Histogram => histograms.remove(histogram)
            case meter: Meter => meters.remove(meter)
            case _ => None
        }

        if (rp.isDefined) {

            val value = rp.get
            val vars = value.group.getAllVariables.asScala.toSeq
            vars.foreach { case (k,v) => dropVar(k) }

            byKey.remove(value.key)

            if (logger.isDebugEnabled()) {
                val id = value.group.getMetricIdentifier(value.name)
                logger.debug(s"DropMetric, Key: ${value.key}, Vars: ${value.group.getAllVariables.size}, Class: ${value.metric.getClass.getName}, ID: ${id}")

            }

        } else if (logger.isDebugEnabled) logger.debug(s"DropMetric, Not Found: (${metric.getClass.getName}, ${metric})")
    }

    override def metrics[C <: ReportContext](context : C) : Seq[context.formatter.Val] = {

        val list = collection.mutable.ListBuffer.empty[context.formatter.Val]

        list ++= counters.map { case (m,v) => formatCounter(v.key, m)(context.formatter) }

        list ++= gauges.map { case(m,v) => formatGauge(v.key, m)(context.formatter) }

        list ++= histograms.map { case (m,v) => formatHistogram(v.key, m)(context.formatter) }.flatMap { i => i }

        list ++= meters.map { case(m,v) => formatMeter(v.key, m)(context.formatter) }.flatMap { i => i}

        val result = list.toList

        if (logger.isDebugEnabled) {
            logger.debug(s"Metrics, Count: ${result.size}, Context: ${context.id}")

            if (logger.isTraceEnabled) logger.trace(s"Values(${result.size}): ${result}, Context: ${context.id}")
        }


        result
    }

    override def vars[C <: ReportContext](context : C) : Map[String, String] = {
        val result = vars.map { case (k,v) => k -> v.value }.toMap

        if (logger.isDebugEnabled()) {
            logger.debug(s"Vars, Count: ${result.size}, Context: ${context.id}")

            if (logger.isTraceEnabled) logger.trace(s"Vars(${result.size}): ${result}, Context: ${context.id}")
        }

        result
    }
}
